src/librbd/ImageCtx.h

```diff
@@ -233,6 +234,8 @@ namespace librbd {
 
     ZTracer::Endpoint trace_endpoint;
 
+    crypto::CryptoInterface* crypto = nullptr;
```

Comment:
Nit: maybe leave this as a `ceph::ref_t` instead of a raw pointer to avoid potential object lifetime races

Reply:
This is what I initially tried and then I got compile errors.
I think that maybe if I add include to CryptoInterface.h from ImageCtx.h it may solve it.
Can you see the issue here?

Reply:
True, I suspect it's because it doesn't know that it derives from `RefCountedObject`. Given that `CryptoInterface` is not very lightweight, I guess it would be better to leave it as a pointer but that means that the shut down state machine should probably grab a reference while it's running.

Reply:
`ShutDownCryptoRequest` doesn't care about the actual `CryptoInterface` object. It might as well get deleted while the state machine is running. Do you agree?
On the other hand, in the future `FormatRequest` may have to grab a reference of `CryptoInterface` while it is serializing it.

Reply:
I'm just looking to avoid the potential for dereferencing an object that is being deleted if a user started two concurrent operations (not that they should be doing that). Since the ref-counting should be responsible for the deletion, I just wanted to keep it consistent.

Reply:
So you want to call `m_image_ctx->crypto.get()` on `ShutDownCryptoRequest<I>::send` and `m_image_ctx->crypto.put()` on `ShutDownCryptoRequest<I>::finish`?

Reply:
Yes please